util/fpga/bitstream_bisect.py

```diff
@@ -0,0 +1,406 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+r"""Efficiently bisect FPGA regressions with the bitstream cache.
+
+This tool bisects while minimizing the number of full bitstream builds. Under
+the hood, it actually runs up to two git bisects. The first bisect only tests
+commits that have cached bitstreams. The optional second bisect narrows down
+ambiguous results. (For more information on bisecting, see `man 1 git-bisect` or
+<https://git-scm.com/docs/git-bisect>.)
+
+In the first phase, the tool classifies commits as good/bad/skip with the
+user-provided `--fast-command`. In the second phase, it uses `--slow-command`.
+These commands classify the commit via their exit status. Exiting with 0
+indicates that a commit is good. Exiting with 125 indicates the commit cannot be
+tested. Exiting with any other status indicates that the commit is bad.
+
+Run bitstream_bisect.py with Bazel like so:
+
+./bazelisk.sh run //util/fpga:bitstream_bisect -- \
+    --good HEAD~30 \
+    --bad HEAD \
+    --fast-command "./bazelisk.sh test //sw/device/tests:uart_smoketest_fpga_cw310_rom" \
+    --slow-command "./bazelisk.sh test --define bitstream=vivado \
+        //sw/device/tests:uart_smoketest_fpga_cw310_rom"
+
+"""
+
+import abc
+import enum
+import os
+import re
+import shlex
+import string
+import subprocess
+from typing import Dict, List, Optional, Set, Tuple
+
+import rich
+import rich.console
+import typer
+
+from rules.scripts.bitstreams_workspace import BitstreamCache
+
+
+class CommitHash(str):
```

Comment:
Nit: It looks like you've got a CommitHash class where "ABC" != "abc"

Reply:
True. I think this is a an unavoidable quirk unless we want to run `git` to canonicalize each value. Another example of the incorrect equality relation is that `abcd` might be short for `abcdef`.

Reply:
If it's worth addressing it can be addressed in a followup. That could check that the hex is equivalent and could accept abbreviations as equal.